### DISCUSSION TITLE
use of printf and serial.printf in libraries - we could replace them externally

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#19356] [Ideas]
Hi All,
I noticed that the rtsp library was emitting serial data (because I had a disconnected USB serial device, and it was echoing back in as serial commands).
These seem to be coming from printf() called within the library. (and maybe Serial.printf() )
This is probably not great for production.... (e.g. in this case, the serial slowing down the udp emission, etc. - so one udp failure could lead to a cascade).

When working on openbeken, I found a way to replace base library functions (specifically printf, vsprintf, as they were disturbingly not thread safe on that build environment, and memory management functions that we wanted to modify).

I wonder if a similar approach may be useful in TAS, replacing printf() with a wrapper which calls AddLog?

The specific modifications were in the makefile:
```
CFLAGS += -DWRAP_PRINTF=1
LFLAGS += -Wl,-wrap,vsnprintf
LFLAGS += -Wl,-wrap,snprintf
LFLAGS += -Wl,-wrap,sprintf
LFLAGS += -Wl,-wrap,vsprintf
```
then providing new functions:
`int __wrap_vsnprintf(char *buffer, size_t bufsz, const char *fmt, va_list val) { ... }`

from memory, the original library function is renamed (so still usable, just under a new name - which can be found in the map file).

If the extensia toolchain supports this (I think it should), then it could be another useful tool in our box of tricks...

Thoughts?

**Comment by Jason2866** [comment]
Why not start first with removing the spamming parts in Lib rtsp? Every not needed output should be removed in every part of the code. Not saying to consider your suggestion.

**Comment by arendst** [comment]
I prefer jasons solution too. It is less impactive while keeping default functionality.

**Comment by barbudor** [comment]
Patching 'print' to a do-nothing stub still has value as it allows to plug new version of offending libs without having to modify the libs source code at each update

**Comment by btsimonh** [comment]
It's particularly useful to know that the facility exists.  For some libraries, we may not have the source at all (probably TAS does not have that issue? - but for openbk, we were linking propreitary libs).
Yes, @Jason2866 's suggestion would work, but then we have to understand what every library contains in detail, and patch each if and as required, understanding the implications.
e.g. the implication for rtsp is that it could cause a cascade type failure in a failure condition.  Because we're not ever seeing the logs without serial connected (I only spotted it because of my serial loop, which was also a chance 'mistake'), and failure conditions are not meant to normally occur, we'd only get reports of 'it performs badly', not 'I get udp send errors when I overload it'.

But the aim of this post is fulfilled.  The idea has lodged as 'possible' in a few key people's minds :).  It's not something we 'must do', certainly not now, but may be useful in the future, or a sunday afternoon project to check it out....

**Comment by arendst** [comment]
As tasmota adds it’s libraries it’s vert simple to search for print(f). With previous libs we commented them out. That’s one of the reasons we only use our own certified libraries instead of using default arduino pushed latest libraries.

**Comment by arendst** [comment]
Just removed debug data from rtsp https://github.com/arendst/Tasmota/commit/0acd5335d52e24eb319eb029f3ea22b5d651b1c0
